Add development instructions and other improvements

mkhorton opened the pull request:
Please do not merge yet. This pull request will be to add development instructions to the README and also additional code improvements, including for PEP8 compliance and to improve the tests.

mkhorton commented:
Hi @stepan-tsirkin, I'm working on the tests for this. Can I ask (1) if the `wannier-spin/run.sh` example still works with the latest code, and (2) what the appropriate `run.sh` commands are for the `CuBi2O4-scalar`, `CuBi2O4-spinor` and `Bi-hoti` examples are?

MIraola commented:
Hi @mkhorton, I found a bug in the interface of Wannier90, maybe it was the source of the problem you point out about the example. At least, not it works fine.
I have added the run.sh commands of all examples and updated their outputs. 
What is the status of this PR? Thanks.

stepan-tsirkin commented:
Ups, I really forgot to reply to this question, sorry for that. Did you find an answer already? Thank you @MIraola  for fixing a bug. @mkhorton , really, what is the status and if there are problems now that prevent to go further?

mkhorton commented:
Hi @MIraola, with the new run.sh examples I can resume work on this PR, I didn't want to change too much unless I could verify with the tests that output was not changing. The first step will be to incorporate the new run.sh examples into the testing, then I can resume making the PEP8 changes.

mkhorton commented:
Also, thank you for adding these!

mkhorton commented:
Ok, I can be more specific about the test failures I was seeing. For example, in the Bi-scalar example, it is simply the case that the order of the detect symmetry operations changes, which is harmless -- the detected irreps are the same. I'll think about if I can structure the test better.

mkhorton commented:
Ok, @stepan-tsirkin, @MIraola, this should be ready to merge unless you have any comments or concerns.

Additional changes:
* I've modified the existing tests to make them more general, see [this file](https://github.com/mkhorton/irrep/blob/master/irrep/tests/test_cli.py) to see what is tested
* I've added a test for the `Bi-hoti` example
* I've added a GitHub Action to automatically run tests on modification, so we will be notified if future code updates break existing examples -- you can see an example of this [here](https://github.com/mkhorton/irrep/actions)
* I have also added a [test badge](https://github.com/mkhorton/irrep/workflows/tests/badge.svg) like this to the README to monitor whether tests have passed
* I've made numerous improvements to follow PEP8 code formatting guidelines and removed unused imports

There are still some additional improvements I could make but these can be done in a separate PR.

stepan-tsirkin reviewed (changes requested):
Hi Matthew, thank you very much for cleaning the code and implementing tests.  
I just have some comments on the pep8 formatting.  I am not a fan of pep8 standard (and of standards in general) but I understand that they may be useful sometimes. However I do not see a goal of multiplying the number of lines in the code so much (unless the developer is payed for every line of course, but we make a free code unfortunately)  like breaking a call to a function of 2 arguments into 3 lines, or making linebreaks inside array indexing. Is it intended or a bug in the conversion script?   

I know, it is a matter of personal taste, but do at least you personally like it. I pointed some places by comments, but I did not look through all files, maybe there are such places more. 

Of course, I am  speaking only about certain places in the code, generally the readability has improved a lot.   What do you think of this? @MIraola  , what is your opinion?

mkhorton commented on irrep/__gvectors.py:
```diff
@@ -15,150 +16,193 @@
-    npw1=igall.shape[1]
-    multZ=np.exp(-1.j*(2*np.pi*A.dot(T).dot(igall[:3,:]+K[:,None])))
-    igrot=transformed_g(K,igall,RecLattice,A)
+        return np.dot(WF[:,
+                         igrot].conj() * WF[:,
+                                            :npw1] * S[0,
+                                                       0] + WF[:,
+                                                               igrot + npw1].conj() * WF[:,
+                                                                                         npw1:] * S[1,
+                                                                                                    1] + WF[:,
+                                                                                                            igrot].conj() * WF[:,
+                                                                                                                               npw1:] * S[0,
+                                                                                                                                          1] + WF[:,
+                                                                                                                                                  igrot + npw1].conj() * WF[:,
+                                                                                                                                                                            :npw1] * S[1,
+                                                                                                                                                                                       0],
+                      multZ)
```
I agree this looks ugly -- the problem is the line length here, there is a way to re-write this and have it still be PEP8 compliant though.

mkhorton replied:
I've just pushed an update to improve this, what do you think of the new version? I think it's less ugly now.

I was initially trying for correctness as the first priority, and then additional improvements later. I would strongly advocate for PEP8 in general.

mkhorton commented on irrep/bandstructure.py:
```diff
@@ -16,550 +16,752 @@
+                LC in zip(
+                    EN,
+                    loc))
+        else:
```
Fixed

mkhorton replied:
To emphasize the above; this is definitely in error. PEP8 should improve readability of code, not decrease it. The tool I was using for a first pass at the above was tripped up by the long line length.

mkhorton commented on irrep/kpoint.py:
```diff
@@ -18,559 +19,732 @@
+                    np.eye(
+                        IBend -
+                        IBstart))) < 1e-10  # check orthonormality
+
```
Fixed

mkhorton commented on irrep/bandstructure.py:
```diff
@@ -16,550 +16,752 @@
+        self.RecLattice = np.array([np.cross(self.Lattice[(i + 1) %
+                                                          3], self.Lattice[(i + 2) %
+                                                                           3]) for i in range(3)]) * 2 * np.pi / np.linalg.det(self.Lattice)
+
```
Line breaks in themselves are ok. This specific one is a bit overzealous; I've also fixed this with the help of the black formatter (mentioned previously).

There are still a few PEP8 violations in this file however, e.g. `self.RecLattice` should be `self.rec_lattice`. I have not fixed all PEP8 issues yet, this PR is just a start.

MIraola commented:
My opinion is similar to yours: I am not used to write code in PEP8 (at least yet), but I agree to the idea that it may be beneficial to respect it, as long as it does not require writing lines that make the code "legal with respect to PEP8" but difficult to follow, like line breaks in indexing of arrays... I believe that it is important to find an equilibrium between readability (PEP8) and comfort of developers (personal tastes).

Generally, I think that the readability of the code has improved a lot.

stepan-tsirkin commented:
Thank you Matthew, Now it is much better,
I see some places that might be improved, but maybe we do it later, by separate PRs, taking advantage of running tests.
